app/config-utils.js

```diff
@@ -0,0 +1,211 @@
+  for (const [k, v] of Object.entries(config)) {
+    switch (k) {
+      case 'federated_extensions':
+        memo[k] = [...(memo[k] || []), ...v];
+        break;
+      default:
+        memo[k] = v;
+    }
+  }
+  return memo;
+}
+
+/**
+ * Load jupyter config data from (this) page and merge with
+ * `jupyter-lite.json#jupyter-config-data`
+ */
+async function getPathConfig(url) {
+  let promises = [getPageConfig(url)];
+  for (const fileName of LITE_FILES) {
+    promises.unshift(getLiteConfig(url, fileName));
+  }
+  return Promise.all(promises);
+}
+
+/**
+ * The current normalized location
+ */
+function here() {
+  return window.location.href.replace(/(\/|\/index.html)?$/, '/');
+}
+
+/**
+ * Maybe fetch an `index.html` in this folder, which must contain the trailing slash.
+ */
+export async function getPageConfig(url = null) {
+  let script = CONFIG_SCRIPT;
+
+  if (url != null) {
+    const text = await (await window.fetch(`${url}index.html`)).text();
+    const doc = parser.parseFromString(text, 'text/html');
+    script = doc.getElementById(JUPYTER_CONFIG_ID);
+  }
+  return fixRelativeUrls(url, JSON.parse(script.textContent));
+}
+
+/**
+ * Fetch a jupyter-lite JSON or Notebook in this folder, which must contain the trailing slash.
+ */
+export async function getLiteConfig(url, fileName) {
+  let text = '{}';
+  let config = {};
+  const liteUrl = `${url || HERE}${fileName}`;
+  try {
+    text = await (await window.fetch(liteUrl)).text();
+    const json = JSON.parse(text);
+    const liteConfig = fileName.endsWith('.ipynb')
+      ? json['metadata']['jupyter-lite']
+      : json;
+    config = liteConfig[JUPYTER_CONFIG_ID] || {};
+  } catch (err) {
+    console.warn(`failed get ${JUPYTER_CONFIG_ID} from ${liteUrl}`);
+  }
+  return fixRelativeUrls(url, config);
+}
+
+export function fixRelativeUrls(url, config) {
+  let urlBase = new URL(url || here()).pathname;
+  for (const [k, v] of Object.entries(config)) {
+    if (k.endsWith('Url') && v.startsWith('./')) {
+      config[k] = `${urlBase}${v.slice(2)}`;
+    }
+  }
+  return config;
+}
+
+/**
+ * Update with the as-configured favicon
+ */
+function addFavicon(config) {
+  const favicon = document.createElement('link');
+  favicon.rel = 'icon';
+  favicon.type = 'image/x-icon';
+  favicon.href = config.faviconUrl;
+  document.head.appendChild(favicon);
+}
+
+/**
+ * The main entry point.
+ */
+async function main() {
+  const config = await jupyterConfigData();
+  if (config.baseUrl === new URL(here()).pathname) {
+    window.location.href = config.appUrl;
+    return;
+  }
+  // rewrite the config
+  CONFIG_SCRIPT.textContent = JSON.stringify(config, null, 2);
+  addFavicon(config);
+  const preloader = document.getElementById(LITE_MAIN);
+  const bundle = document.createElement('script');
+  bundle.src = preloader.href;
+  bundle.main = preloader.main;
+  document.head.appendChild(bundle);
+}
+
+/**
+ * TODO: consider better pattern for invocation.
+ */
+await main();
```

Comment:
Looks like this might give a top level await error on some browsers.

Reply:
yeah, i just wanted something that _didn't_ webpack for this stage. as mentioned, we likely want this for "in-app" stuff too, so it probably needs to become a package with an entry point in the top-level webpack.

We also need to do more work on cache-busting, getting that `[hash]` back into file names, as I've noted on RTD it's pretty easy to get in a bad cache state. there's probably some more stuff we can fiddle with to get more-reproducible chunks.

Reply:
So top level await seems to be supported by Chromium 90 and Firefox 89.0b7 (developer edition). But not in Firefox 88.

Reply:
Actually do we need the await here? Or could this be changed to the following (at least for now):

```typescript
void main();
```